### PR TITLE
[FLINK-21242][python] Support state access API for the map/flat_map/filter/reduce operations of Python KeyedStream

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request supports state access API for the map/flat_map/filter/reduce operations of Python KeyedStream.*

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-python/pyflink/datastream/data_stream.py [hunk 1]
```diff
@@ -785,11 +785,49 @@ def __init__(self, j_keyed_stream, original_data_type_info, origin_stream: DataS
...
+        return self.process(KeyedMapFunctionWrapper(func), output_type)  # type: ignore
 
     def flat_map(self, func: Union[Callable, FlatMapFunction], result_type: TypeInformation = None)\
             -> 'DataStream':
-        return self._values().flat_map(func, result_type)
+        if not isinstance(func, FlatMapFunction):
+            if callable(func):
+                func = FlatMapFunctionWrapper(func)  # type: ignore
+            else:
+                raise TypeError("The input must be a FlatMapFunction or a callable function")
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/data_stream.py:815]:
```suggestion
                raise TypeError("The input func must be a FlatMapFunction or a callable function")
```

File: flink-python/pyflink/datastream/data_stream.py [hunk 2]
```diff
@@ -785,11 +785,49 @@ def __init__(self, j_keyed_stream, original_data_type_info, origin_stream: DataS
 
     def map(self, func: Union[Callable, MapFunction], output_type: TypeInformation = None) \
             -> 'DataStream':
-        return self._values().map(func, output_type)
+        if not isinstance(func, MapFunction):
+            if callable(func):
+                func = MapFunctionWrapper(func)  # type: ignore
+            else:
+                raise TypeError("The input must be a MapFunction or a callable function")
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/data_stream.py:792]:
```suggestion
                raise TypeError("The input func must be a MapFunction or a callable function")
```

File: flink-python/pyflink/datastream/data_stream.py [hunk 3]
```diff
@@ -811,19 +849,55 @@ def reduce(self, func: Union[Callable, ReduceFunction]) -> 'DataStream':
...
+                return [reduce_value]
+
+        return self.process(KeyedReduceFunctionWrapper(func), output_type)  # type: ignore
 
     def filter(self, func: Union[Callable, FilterFunction]) -> 'DataStream':
-        return self._values().filter(func)
+        if callable(func):
+            func = FilterFunctionWrapper(func)  # type: ignore
+        elif not isinstance(func, FilterFunction):
+            raise TypeError("func must be a Callable or instance of FilterFunction.")
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/data_stream.py:881]:
```suggestion
            raise TypeError("The input func must be a FilterFunction or a callable function!")
```

**Review by HuangXingBo — CHANGES_REQUESTED** [review]
@WeiZhong94 Thanks a lot for the PR. Looks good overall. I only left some minor comments.

**Comment by WeiZhong94** [comment]
@HuangXingBo Thanks for your review! I have fixed the test failure and addressed your comments.

**Review by HuangXingBo — APPROVED** [review]
@WeiZhong94 Thanks a lot for the update. LGTM. The failed test has been recorded in [FLINK-21647](https://issues.apache.org/jira/browse/FLINK-21647) and it seems to has been resolved in [FLINK-21128](https://issues.apache.org/jira/browse/FLINK-21128). I think you can rebase the`master` and trigger an Azure build again.

**Comment by WeiZhong94** [comment]
@flinkbot run azure

**Pull request merged**